build/Shared/SharedExtensions.cs

```diff
@@ -126,5 +125,49 @@ public static void AppendInt(this StringBuilder sb, int value)
                 divisor /= 10;
             }
         }
+
+        public static int SingleIndex<T>(this IEnumerable<T> lst, Predicate<T> isMatch)
```

Comment:
Given that this method is so far invoked on `PackageSpec.TargetFrameworks` which is of type `List`, I think we could reduce the allocations here by allowing this method to be invoked on the `IList` and then invoke `lst.NoAllocEnumerate` in the forearch statement.

https://github.com/NuGet/NuGet.Client/blob/8e96cf5c2b5a192a3955b7e36951e482b2613915/build/Shared/NoAllocEnumerateExtensions.cs#L34-L46